orttraining/orttraining/training_ops/cuda/nn/dropout_impl.cu

```diff
@@ -24,15 +24,22 @@
 namespace onnxruntime {
 namespace cuda {
 
-template <typename T>
+template <typename T, int NumThreadsPerBlock, int NumElementsPerThread>
```

Comment:
>NumThreadsPerBlock [](start = 26, length = 18)

is this equivalent to blockDim?

Reply:
yes. NumThreadsPerBlock and NumElementsPerThread are passed in as static compile time variable so that #pragma unroll can optimize them away 

---
In reply to: [441079699](https://github.com/microsoft/onnxruntime/pull/4167#discussion_r441079699) [](ancestors = 441079699)

Reply:
i see how NumElementsPerThread can be used for unrolling, how does NumThreadsPerBlock affect it?

Reply:
It's not strictly needed... one can also use blocksize .... but I am following the convention in other elementwise kernels. 

---
In reply to: [443858656](https://github.com/microsoft/onnxruntime/pull/4167#discussion_r443858656) [](ancestors = 443858656)

Reply:
sometimes conventions can be improved :)
it seems redundant to specify it twice when launching the kernel. i'm fine with leaving it as is though

Reply:
Let's do it another PR to fix the convention. 

---
In reply to: [444609787](https://github.com/microsoft/onnxruntime/pull/4167#discussion_r444609787) [](ancestors = 444609787)

Reply:
It is different between blocksize and template parameter. NumThreadsPerBlock  and NumElementsPerThread both are constant, so compiler can do more optimization. But the value of blocksize is unknown during compiling.